### PR TITLE
Hive: Make HiveCatalog based tables automatically readable from Hive

### PR DESCRIPTION
**Pull request by pvary**
- Added the needed extra fields on table creation.
- Updated the TestHiveIcebergStorageHandlerWithHiveCatalog tests so they do not make the same modifications by themselves.

After the changes the tests are still able to queries, so we should be happy

### CODE REVIEW

**Comment by rdblue** [comment]
The changes look good. Just a quick sanity check: what happens in Hive if the serde or storage handler class is missing? Will this break metastore operations on the table?

I remember that a missing class used to break any query that loaded the table, including DROP TABLE. So you couldn't even clean up problems.

**Comment by pvary** [comment]
> The changes look good. Just a quick sanity check: what happens in Hive if the serde or storage handler class is missing? Will this break metastore operations on the table?
> 
> I remember that a missing class used to break any query that loaded the table, including DROP TABLE. So you couldn't even clean up problems.

That was my fear as well, but did not have time to test them out yet. It might even cause issues for users who are not using the Hive tables, just want to use HiveCatalog to store the table versions.

**Comment by pvary** [comment]
> That was my fear as well, but did not have time to test them out yet. It might even cause issues for users who are not using the Hive tables, just want to use HiveCatalog to store the table versions.

Your comment give me a better idea how to test. Changed the serde to an arbitrary string, and got this Exception:
```
FAILED: RuntimeException org.apache.hadoop.hive.ql.metadata.HiveException: Error in loading storage handler.org.apache.iceberg.mr.hive.HiveIcebergStorageHandler3
[a3db973d-5b1f-4fde-985b-ad5aa4d09018 Thread-4] ERROR org.apache.hadoop.hive.ql.Driver - FAILED: RuntimeException org.apache.hadoop.hive.ql.metadata.HiveException: Error in loading storage handler.org.apache.iceberg.mr.hive.HiveIcebergStorageHandler3
java.lang.RuntimeException: org.apache.hadoop.hive.ql.metadata.HiveException: Error in loading storage handler.org.apache.iceberg.mr.hive.HiveIcebergStorageHandler3
	at org.apache.hadoop.hive.ql.metadata.Table.getStorageHandler(Table.java:297)
	at org.apache.hadoop.hive.ql.metadata.Table.getInputFormatClass(Table.java:307)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.getMetaData(SemanticAnalyzer.java:2041)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.getMetaData(SemanticAnalyzer.java:1934)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.genResolvedParseTree(SemanticAnalyzer.java:11080)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.analyzeInternal(SemanticAnalyzer.java:11133)
	at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.analyzeInternal(SemanticAnalyzer.java:10807)
	at org.apache.hadoop.hive.ql.parse.BaseSemanticAnalyzer.analyze(BaseSemanticAnalyzer.java:258)
	at org.apache.hadoop.hive.ql.Driver.compile(Driver.java:512)
	at org.apache.hadoop.hive.ql.Driver.compileInternal(Driver.java:1317)
	at org.apache.hadoop.hive.ql.Driver.compileAndRespond(Driver.java:1295)
	at org.apache.hive.service.cli.operation.SQLOperation.prepare(SQLOperation.java:204)
	at org.apache.hive.service.cli.operation.SQLOperation.runInternal(SQLOperation.java:290)
	at org.apache.hive.service.cli.operation.Operation.run(Operation.java:320)
	at org.apache.hive.service.cli.session.HiveSessionImpl.executeStatementInternal(HiveSessionImpl.java:530)
	at org.apache.hive.service.cli.session.HiveSessionImpl.executeStatement(HiveSessionImpl.java:500)
	at org.apache.hive.service.cli.CLIService.executeStatement(CLIService.java:265)
	at com.klarna.hiverunner.HiveServerContainer.executeStatement(HiveServerContainer.java:120)
	at com.klarna.hiverunner.builder.HiveShellBase.executeStatementsWithCommandShellEmulation(HiveShellBase.java:115)
	at com.klarna.hiverunner.builder.HiveShellBase.executeStatementWithCommandShellEmulation(HiveShellBase.java:109)
	at com.klarna.hiverunner.builder.HiveShellBase.executeStatement(HiveShellBase.java:99)
	at org.apache.iceberg.mr.hive.HiveIcebergStorageHandlerBaseTest.testScanEmptyTable(HiveIcebergStorageHandlerBaseTest.java:119)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at com.klarna.hiverunner.StandaloneHiveRunner.evaluateStatement(StandaloneHiveRunner.java:165)
	at com.klarna.hiverunner.HiveRunnerRule$HiveRunnerRuleStatement.evaluate(HiveRunnerRule.java:82)
	at com.klarna.hiverunner.ThrowOnTimeout$1.run(ThrowOnTimeout.java:60)
	at java.lang.Thread.run(Thread.java:748)
```

Is it ok to require "iceberg-mr" classes on the classpath to use HiveCatalog?

I would be afraid of this requirement until we do not have a Catalog implementation which could serve as a replacement catalog in case we do not have HDFS at hand.

**Comment by massdosage** [comment]
> > The changes look good. Just a quick sanity check: what happens in Hive if the serde or storage handler class is missing? Will this break metastore operations on the table?
> > I remember that a missing class used to break any query that loaded the table, including DROP TABLE. So you couldn't even clean up problems.
> 
> That was my fear as well, but did not have time to test them out yet. It might even cause issues for users who are not using the Hive tables, just want to use HiveCatalog to store the table versions.

I'm pretty sure that if you try to perform any operations on tables with the StorageHandler set on them, and you don't have that class on the classpath, it will fail with an error. I've been in exactly the situation you describe where I couldn't issue a DROP TABLE for a table that was created by a test storage handler that no longer existed.

**Comment by pvary** [comment]
> > > The changes look good. Just a quick sanity check: what happens in Hive if the serde or storage handler class is missing? Will this break metastore operations on the table?
> > > I remember that a missing class used to break any query that loaded the table, including DROP TABLE. So you couldn't even clean up problems.
> > 
> > 
> > That was my fear as well, but did not have time to test them out yet. It might even cause issues for users who are not using the Hive tables, just want to use HiveCatalog to store the table versions.
> 
> I'm pretty sure that if you try to perform any operations on tables with the StorageHandler set on them, and you don't have that class on the classpath, it will fail with an error. I've been in exactly the situation you describe where I couldn't issue a DROP TABLE for a table that was created by a test storage handler that no longer existed.

The tests caught the exact same problem:
```
org.apache.iceberg.spark.sql.TestAlterTable > testSetTableProperties[2] FAILED
    java.lang.RuntimeException: org.apache.hadoop.hive.ql.metadata.HiveException: Error in loading storage handler.org.apache.iceberg.mr.hive.HiveIcebergStorageHandler
        at org.apache.hadoop.hive.ql.metadata.Table.getStorageHandler(Table.java:297)
        at org.apache.spark.sql.hive.client.HiveClientImpl.convertHiveTableToCatalogTable(HiveClientImpl.scala:465)
        at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$getTableOption$3(HiveClientImpl.scala:424)
        at scala.Option.map(Option.scala:230)
        at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$getTableOption$1(HiveClientImpl.scala:424)
```

My first ideas:
1. Check that the classes are on the classpath before creating the table? - This would only fix the problem if the creator/reader of the table is both have the same things on the classpath.
2. Move the HiveIcebergStorageHandler / HiveIcebergSerDe to their own package and try to shave of the dependencies - Seems like a hard solution and we will eventually fall back to Reflection
3. Go full reflection: Create a reflection based SerDe, StorageHandler which checks the classpath for the actual classes and if successful wraps them?

@rdblue, @massdosage: What do you think about the 3rd option?

**Comment by rdblue** [comment]
Would option 3 work? The reflection-based loader would need to be in the Hive classpath at all times. So we would have the same problem with that one right? This is a problem that Hive would need to fix.

I lean toward the approach of making it optional to set the serde and storage handler, and using either reflection to see if the classes currently exist or a table property to change the behavior. A table property makes sense to me: `engine.hive.enabled=true` could signal that these should be real classes.

What do you think?

**Comment by pvary** [comment]
> Would option 3 work? The reflection-based loader would need to be in the Hive classpath at all times. So we would have the same problem with that one right? This is a problem that Hive would need to fix.

Without the HiveIcebergStorageHandler, Hive will not even know the columns in the table (some historic version is stored in HMS, but not considered correct and queried again from the StorageHandler every time). So Hive considers these tables corrupt. Not being able to drop them is definitely a problem, but for the other functions it is questionable if we provide any data / possibility to manipulate.

> I lean toward the approach of making it optional to set the serde and storage handler, and using either reflection to see if the classes currently exist or a table property to change the behavior. A table property makes sense to me: `engine.hive.enabled=true` could signal that these should be real classes.

Config still would mean that if we turn that on for a table then we need the HiveIcebergStorageHandler on the classpath for even the Spark processes which are accessing the table.

In my opinion this is really a static/dynamic decision. With config we statically set a table to Hive or Spark readable one. With the reflection based solution we allow the accessor to dynamically decide which StorageHandler is needed. But you are absolutely right that if Hive access is required, then something has to be on the classpath.

Would combining the 2 be an overkill?
Like config if you want or not, and if you want ReflectionStorageHandler would be set. So if enabled the Spark could still access the HiveCatalog without the StorageHandler, and Hive will use HiveIcebergStorageHandler.

I see the following uses-cases:
1. Users have data on S3 but need some consistent store on the cloud. No HDFS at hand, so they need HiveCatalog, but really uses only Spark
2. Users use Spark and only occasionally Hive. Since they already have Hive at hand, uses HiveCatalog
3. Users use Spark and Hive too. HiveCatalog is an obvious choice
4. Users use Hive but want to use Iceberg for the partitioning and such

You might know more about other use-cases to consider...

**Comment by rdblue** [comment]
I don't think the reflection option would have any benefit because Hive would still need `ReflectionStorageHandler` in the classpath. If we can guarantee that something is in the classpath, then we should use `HiveIcebergStorageHandler` so everything works.

Then we have the choice between whether we break Hive or not. I think a config property is the right way to go so that it is up to the user or administrator whether to expect Iceberg in the classpath. I think maybe a Hadoop Configuration option in hive-site.xml is a good idea.

**Comment by pvary** [comment]
> I don't think the reflection option would have any benefit because Hive would still need `ReflectionStorageHandler` in the classpath.

This is absolutely true for Hive!
I am a little bit concerned about the Spark packages. We might be better of without the iceberg-hive-mr package classes in the iceberg-spark-runtime libraries. The HiveIcebergStorageHandler would be needed only to be present there and would not be used at all, like every other class in the iceberg-hive-mr package.

If you think that extra few classes is an acceptable compromise to have less complexity, then I would be happy to move forward with the configuration solution.

**Comment by pvary** [comment]
> Then we have the choice between whether we break Hive or not. I think a config property is the right way to go so that it is up to the user or administrator whether to expect Iceberg in the classpath. I think maybe a Hadoop Configuration option in hive-site.xml is a good idea.

Implemented the config property + hive-site.xml solution.
@rdblue: Could you please check that this is the same solution you were thinking about?
Also please double check the dependency changes I have added.
Thanks,
Peter

**Comment by rdblue** [comment]
Okay, I understand what you're saying about `ReflectionStorageHandler` now (for Spark). But I don't think we need it because the Spark code interacts with the table through the Iceberg library. It won't try to instantiate the storage handler because that's specific to the Hive integraiton.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 1]
```diff
@@ -169,6 +171,15 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
         tbl.getParameters().put("EXTERNAL", "TRUE"); // using the external table type also requires this
       }
 
+      // If  needed set the required properties on the table to be able to query from Hive
+      if (hiveEngineEnabled(metadata, conf)) {
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:175]: I think this should be in the `storageDescriptor` method instead of here in the commit.

File: core/src/main/java/org/apache/iceberg/TableProperties.java [hunk 2]
```diff
@@ -126,4 +126,8 @@ private TableProperties() {
 
   public static final String SNAPSHOT_ID_INHERITANCE_ENABLED = "compatibility.snapshot-id-inheritance.enabled";
   public static final boolean SNAPSHOT_ID_INHERITANCE_ENABLED_DEFAULT = false;
+
+  public static final String ENGINE_HIVE_ENABLED = "engine.hive.enabled";
+  public static final String ENGINE_HIVE_ENABLED_HIVE_CONF = "iceberg.engine.hive.enabled";
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/TableProperties.java:131]: Could we move this to a class for Hadoop configuration properties, like `org.apache.iceberg.hadoop.ConfigProperties`?

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableTest.java [hunk 3]
```diff
@@ -325,4 +329,71 @@ public void testNonDefaultDatabaseLocation() throws IOException, TException {
...
+  @Test
+  public void testEngineHiveEnabled() throws TException {
+    // Default already created
+    org.apache.hadoop.hive.metastore.api.Table hmsTable = metastoreClient.getTable(DB_NAME, TABLE_NAME);
+
+    assertHiveEnabled(hmsTable, false);
+
+    catalog.dropTable(TABLE_IDENTIFIER);
+
+    // Enable by hive-conf
```
Comment (rdblue) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableTest.java:341]: It looks like a lot of test cases are mixed together in this method, but separated by comments. I think it would be better to make these separate test methods because then they are run independently, so you can see all of the failures at once, instead of just one failure per run that potentially hides the others.

**Comment by rdblue** [comment]
The approach looks good to me. Thanks @pvary!

Should we also add this to table property documentation?

@massdosage, you might want to have a look as well. This sets the Hive storage handler and serde if a hive-site.xml property is set, or if the table has a configuration option.

File: core/src/main/java/org/apache/iceberg/TableProperties.java (hunk 2 above)
Comment (pvary) [reply core/src/main/java/org/apache/iceberg/TableProperties.java]: Done

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableTest.java (hunk 3 above)
Comment (pvary) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableTest.java:346]: Done (kept true/false tests in a single test case, but separated on config source)

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 1 above)
Comment (pvary) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java]: Done

**Comment by pvary** [comment]
> Okay, I understand what you're saying about `ReflectionStorageHandler` now (for Spark). But I don't think we need it because the Spark code interacts with the table through the Iceberg library. It won't try to instantiate the storage handler because that's specific to the Hive integraiton.

We need the build changes, because the spark sql uses Hive code to interact with hive tables. This fact was highlighted by the failures in the tests.

You can repro by reverting the `gradle.build` changes and setting the `ENGINE_HIVE_ENABLED_DEFAULT` to `true`. 
```
./gradlew :iceberg-spark3:test --tests org.apache.iceberg.spark.sql.TestCreateTable
```

One of the exception what I have got is here:
```
    java.lang.RuntimeException: org.apache.hadoop.hive.ql.metadata.HiveException: Error in loading storage handler.org.apache.iceberg.mr.hive.HiveIcebergStorageHandler
        at org.apache.hadoop.hive.ql.metadata.Table.getStorageHandler(Table.java:297)
        at org.apache.spark.sql.hive.client.HiveClientImpl.convertHiveTableToCatalogTable(HiveClientImpl.scala:465)
        at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$getTableOption$3(HiveClientImpl.scala:424)
        at scala.Option.map(Option.scala:230)
        at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$getTableOption$1(HiveClientImpl.scala:424)
        at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$withHiveState$1(HiveClientImpl.scala:294)
        at org.apache.spark.sql.hive.client.HiveClientImpl.liftedTree1$1(HiveClientImpl.scala:227)
        at org.apache.spark.sql.hive.client.HiveClientImpl.retryLocked(HiveClientImpl.scala:226)
        at org.apache.spark.sql.hive.client.HiveClientImpl.withHiveState(HiveClientImpl.scala:276)
        at org.apache.spark.sql.hive.client.HiveClientImpl.getTableOption(HiveClientImpl.scala:422)
        at org.apache.spark.sql.hive.client.HiveClient.getTable(HiveClient.scala:90)
        at org.apache.spark.sql.hive.client.HiveClient.getTable$(HiveClient.scala:89)
        at org.apache.spark.sql.hive.client.HiveClientImpl.getTable(HiveClientImpl.scala:90)
        at org.apache.spark.sql.hive.HiveExternalCatalog.getRawTable(HiveExternalCatalog.scala:120)
        at org.apache.spark.sql.hive.HiveExternalCatalog.$anonfun$getTable$1(HiveExternalCatalog.scala:719)
        at org.apache.spark.sql.hive.HiveExternalCatalog.withClient(HiveExternalCatalog.scala:99)
        at org.apache.spark.sql.hive.HiveExternalCatalog.getTable(HiveExternalCatalog.scala:719)
        at org.apache.spark.sql.catalyst.catalog.ExternalCatalogWithListener.getTable(ExternalCatalogWithListener.scala:138)
        at org.apache.spark.sql.catalyst.catalog.SessionCatalog.getTableMetadata(SessionCatalog.scala:446)
        at org.apache.spark.sql.execution.command.DropTableCommand.run(ddl.scala:226)
        at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)
        at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)
        at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:79)
        at org.apache.spark.sql.Dataset.$anonfun$logicalPlan$1(Dataset.scala:229)
        at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3616)
        at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:100)
        at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:160)
        at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:87)
        at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:763)
        at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
        at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3614)
        at org.apache.spark.sql.Dataset.<init>(Dataset.scala:229)
        at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:100)
        at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:763)
        at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:97)
        at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:606)
        at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:763)
        at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:601)
        at org.apache.iceberg.spark.SparkTestBase.sql(SparkTestBase.java:83)
        at org.apache.iceberg.spark.sql.TestCreateTable.dropTestTable(TestCreateTable.java:47)
```

**Comment by rdblue** [comment]
> We need the build changes, because the spark sql uses Hive code to interact with hive tables.

That sounds like a bug to me. We want to ensure that Spark does not interact with Iceberg tables using Hive. It should use Iceberg directly through the v2 interface instead. We'll need to find out why Spark is trying to load the table incorrectly. What catalog were you using?

**Comment by pvary** [comment]
> > We need the build changes, because the spark sql uses Hive code to interact with hive tables.
> 
> That sounds like a bug to me. We want to ensure that Spark does not interact with Iceberg tables using Hive. It should use Iceberg directly through the v2 interface instead. We'll need to find out why Spark is trying to load the table incorrectly. What catalog were you using?

I did the following 2 changes in the Iceberg code, and run the spark3 tests:
- Remove the additional dependencies, and 
- Set the default value for  to `false` and run the spark3 tests.

One such test is `TestCreateTable.testCreateTable`. The drop table in the "after" fails with the exception below.

It uses HiveCatalog. See: [code](https://github.com/apache/iceberg/blob/1772f4f27b8a12d3e89a7f65b8b600b717e1f09d/spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java#L63)

It tries to run `DROP TABLE IF EXISTS testhive.default.table` using `spark.sql`. My understanding is that this is spark code, and it tries to delegate it to the embedded `org.apache.hadoop.hive.ql.metadata.Table` class which tries to read the StorageHandler, and fails.

This would mean that when Iceberg tables are using the HiveCatalog and Spark code tries to access them with SparkSQL, it will fail for several commands if we do not have `HiveIcebergStorageHandler` on the classpath.

I am not sure we can change this part of Spark. I think,
- We either have to add a note to the docs that an additional jar is needed if SparkSQL and HiveCatalog is used in conjunction, or 
- We have to add the jars to our runtime jars.

Your thoughts?

The exception stack trace:
```
    java.lang.RuntimeException: org.apache.hadoop.hive.ql.metadata.HiveException: Error in loading storage handler.org.apache.iceberg.mr.hive.HiveIcebergStorageHandler
        at org.apache.hadoop.hive.ql.metadata.Table.getStorageHandler(Table.java:297)
        at org.apache.spark.sql.hive.client.HiveClientImpl.convertHiveTableToCatalogTable(HiveClientImpl.scala:465)
        at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$getTableOption$3(HiveClientImpl.scala:424)
        at scala.Option.map(Option.scala:230)
        at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$getTableOption$1(HiveClientImpl.scala:424)
        at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$withHiveState$1(HiveClientImpl.scala:294)
        at org.apache.spark.sql.hive.client.HiveClientImpl.liftedTree1$1(HiveClientImpl.scala:227)
        at org.apache.spark.sql.hive.client.HiveClientImpl.retryLocked(HiveClientImpl.scala:226)
        at org.apache.spark.sql.hive.client.HiveClientImpl.withHiveState(HiveClientImpl.scala:276)
        at org.apache.spark.sql.hive.client.HiveClientImpl.getTableOption(HiveClientImpl.scala:422)
        at org.apache.spark.sql.hive.client.HiveClient.getTable(HiveClient.scala:90)
        at org.apache.spark.sql.hive.client.HiveClient.getTable$(HiveClient.scala:89)
        at org.apache.spark.sql.hive.client.HiveClientImpl.getTable(HiveClientImpl.scala:90)
        at org.apache.spark.sql.hive.HiveExternalCatalog.getRawTable(HiveExternalCatalog.scala:120)
        at org.apache.spark.sql.hive.HiveExternalCatalog.$anonfun$getTable$1(HiveExternalCatalog.scala:719)
        at org.apache.spark.sql.hive.HiveExternalCatalog.withClient(HiveExternalCatalog.scala:99)
        at org.apache.spark.sql.hive.HiveExternalCatalog.getTable(HiveExternalCatalog.scala:719)
        at org.apache.spark.sql.catalyst.catalog.ExternalCatalogWithListener.getTable(ExternalCatalogWithListener.scala:138)
        at org.apache.spark.sql.catalyst.catalog.SessionCatalog.getTableMetadata(SessionCatalog.scala:446)
        at org.apache.spark.sql.execution.command.DropTableCommand.run(ddl.scala:226)
        at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)
        at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)
        at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:79)
        at org.apache.spark.sql.Dataset.$anonfun$logicalPlan$1(Dataset.scala:229)
        at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3616)
        at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:100)
        at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:160)
        at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:87)
        at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:763)
        at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
        at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3614)
        at org.apache.spark.sql.Dataset.<init>(Dataset.scala:229)
        at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:100)
        at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:763)
        at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:97)
        at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:606)
        at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:763)
        at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:601)
        at org.apache.iceberg.spark.SparkTestBase.sql(SparkTestBase.java:83)
        at org.apache.iceberg.spark.sql.TestCreateTable.dropTestTable(TestCreateTable.java:47)
```

**Comment by rdblue** [comment]
Spark is using a v1 command to drop the table: `org.apache.spark.sql.execution.command.DropTableCommand`. That is unexpected because the test should be using an Iceberg implementation of Spark's `TableCatalog` for all of these operations. Can you find out what catalog was being used? My guess is that it was the built-in [`spark_catalog`](https://github.com/apache/iceberg/blob/1772f4f27b8a12d3e89a7f65b8b600b717e1f09d/spark3/src/test/java/org/apache/iceberg/spark/SparkCatalogTestBase.java#L66-L71) and that somehow the session catalog wrapper that we injected did not correctly detect that this table is Iceberg and not Hive.

**Comment by pvary** [comment]
> Spark is using a v1 command to drop the table: `org.apache.spark.sql.execution.command.DropTableCommand`. That is unexpected because the test should be using an Iceberg implementation of Spark's `TableCatalog` for all of these operations. Can you find out what catalog was being used? My guess is that it was the built-in [`spark_catalog`](https://github.com/apache/iceberg/blob/1772f4f27b8a12d3e89a7f65b8b600b717e1f09d/spark3/src/test/java/org/apache/iceberg/spark/SparkCatalogTestBase.java#L66-L71) and that somehow the session catalog wrapper that we injected did not correctly detect that this table is Iceberg and not Hive.
> 
> So the question is probably why is [`loadTable`](https://github.com/apache/iceberg/blob/1772f4f27b8a12d3e89a7f65b8b600b717e1f09d/spark3/src/main/java/org/apache/iceberg/spark/SparkSessionCatalog.java#L116-L122) catching `NoSuchTableException` and returning the table through Hive?

I did not have too much time to play around with this 😢, and I was not able to catch where the `DROP TABLE` command was hijacked yet.
What I have confirmed:
- It is only happening with `spark_catalog`
- It is only for `DROP TABLE` - I have not found problems with another sql command

What I have done:
- Removed the dependencies

Since this is an unrelated to the original change, we might want to file an Issue for it.

What do you think @rdblue?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 4]
```diff
@@ -169,6 +173,12 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
         tbl.getParameters().put("EXTERNAL", "TRUE"); // using the external table type also requires this
       }
 
+      // If needed set the 'storate_handler' property to enable query from Hive
+      if (hiveEngineEnabled) {
+        tbl.getParameters().put(hive_metastoreConstants.META_TABLE_STORAGE,
+            "org.apache.iceberg.mr.hive.HiveIcebergStorageHandler");
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:179]: Should this be removed if the engine is not enabled? Because this is a table property, I don't think it would be automatically removed at any point.

Comment (pvary) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:179]: Good catch!
Done

**Comment by pvary** [comment]
> > Spark is using a v1 command to drop the table: `org.apache.spark.sql.execution.command.DropTableCommand`. That is unexpected because the test should be using an Iceberg implementation of Spark's `TableCatalog` for all of these operations. Can you find out what catalog was being used? My guess is that it was the built-in [`spark_catalog`](https://github.com/apache/iceberg/blob/1772f4f27b8a12d3e89a7f65b8b600b717e1f09d/spark3/src/test/java/org/apache/iceberg/spark/SparkCatalogTestBase.java#L66-L71) and that somehow the session catalog wrapper that we injected did not correctly detect that this table is Iceberg and not Hive.

I guess this is a spark bug.

What I see is that the spark_catalog is correctly set (CatalogManager.defaultSessionCatalog is the original inner HiveSessionCatalog, but the CatalogManager.catalogs has a value for spark_catalog pointing to our very own SparkSessionCatalog).
Sadly this code does not trigger DropTable command creation during the planning:
```
    case DropTableStatement(NonSessionCatalogAndTable(catalog, tbl), ifExists, _) =>
      DropTable(catalog.asTableCatalog, tbl.asIdentifier, ifExists)
```
https://github.com/apache/spark/blob/v3.0.0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveCatalogs.scala#L175-L176

I know from debugging that the NonSessionCatalogAndTable is None (since with spark_catalog we replace session catalog so it is not a NonSessionCatalogAndTable), so I guess we do not translate the DropTableStatement to DropTable.
We fall back to ResolvaSessionCatalog and convert the statement to DropTableCommand which does not care about the catalog
```
    case DropTableStatement(TempViewOrV1Table(name), ifExists, purge) =>
      DropTableCommand(name.asTableIdentifier, ifExists, isView = false, purge = purge)
```
https://github.com/apache/spark/blob/v3.0.0/sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala#L331-L332

My theory is that we are missing the checks and different code paths for V2 providers which we have for CreateTableStatement for example:
```
      if (!isV2Provider(provider)) {
[..]
      } else {
[..]
      }
```
https://github.com/apache/spark/blob/v3.0.0/sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala#L248

So to summarize the things again:
- This is a Spark bug, where the `ResolveSessionCatalog` should use different code for V1 and V2 providers when a DropTableStatement is issued

@rdblue: I am not an expert in Scala, so happy to hear any objections if I made mistakes. Does my analysis makes sense? How should we proceed?

Thanks,
Peter

**Comment by rdblue** [comment]
If it's a Spark bug, then let's just add an `Assume` for the failing test to exclude it when the catalog is `spark_catalog`. Thanks for looking into it further, @pvary!

**Comment by pvary** [comment]
> If it's a Spark bug, then let's just add an `Assume` for the failing test to exclude it when the catalog is `spark_catalog`. Thanks for looking into it further, @pvary!

Currently in our spark tests we set `engine.hive.enabled` to `false` (and also we default to there so no breaking change in the release). That is way with the code in this PR does not trigger the issue.

So I think we are good on this. Unless we want specific tests in spark sql to have `engine.hive.enabled` set to `true` and test this code path as well.

**Comment by rdblue** [comment]
That sounds fine to me. I thought this was a problem because you were hitting it in tests.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks @pvary! I've merged this. We should also update the docs to explain when the storage handler will be set so that people can update their hive-site.xml if they choose.

**Comment by pvary** [comment]
> Thanks @pvary! I've merged this. We should also update the docs to explain when the storage handler will be set so that people can update their hive-site.xml if they choose.

Am I right when I think this should be in https://github.com/apache/iceberg/blob/master/site/docs/hive.md?

Thanks
Peter

**Comment by massdosage** [comment]
> > Thanks @pvary! I've merged this. We should also update the docs to explain when the storage handler will be set so that people can update their hive-site.xml if they choose.
> 
> Am I right when I think this should be in https://github.com/apache/iceberg/blob/master/site/docs/hive.md?
> 
> Thanks
> Peter

Yes, I was going to add a section under `Using Hadoop Tables`, something like  `Using Hive Catalog` and this was going to be about finding the Hive table created by Iceberg if you use that method and then how to add the StorageHandler to this table. I think we still need something like that (for existing tables and those created without the config changes in this PR) but we also need a section describing how that can automatically be done at Table creation time from this PR and the `hive-site.xml` etc.
